### PR TITLE
proxy/protocol: Add checks on header length and flags

### PR DESCRIPTION
**Pull request by dlespiau**
We can check better the header we receive from clients. This time, make
sure the payload length is within reasonable limits and that the flags
are 0 (we don't have any at the moment).

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

### CODE REVIEW

**Comment by chavafg** [comment]
qa-passed

File: proxy/api/protocol.go [hunk 1]
```diff
@@ -28,6 +29,18 @@ type header struct {
 	flags  uint32
 }
 
+const maxPayloadLength = 1 * 1024 * 1024
```
Comment (jodh-intel) [inline proxy/api/protocol.go]: - What's the `1` for?
- Is there any value in making this a cli option?

File: proxy/api/protocol.go [hunk 2]
```diff
@@ -28,6 +29,18 @@ type header struct {
...
+
+func (hdr *header) validate() error {
```
Comment (jodh-intel) [inline proxy/api/protocol.go]: Could you add a test for this?

File: proxy/api/protocol.go [hunk 3]
```diff
@@ -28,6 +29,18 @@ type header struct {
...
+	if hdr.length > maxPayloadLength {
+		return fmt.Errorf("payload size too big: %d", hdr.length)
```
Comment (jodh-intel) [inline proxy/api/protocol.go]: It would be useful to include `maxPayloadLength` in the error message.

File: proxy/api/protocol.go (hunk 1 above)
Comment (dlespiau) [inline proxy/api/protocol.go]: - that's for 1 MB
- I don't think so

Comment (jodh-intel) [inline proxy/api/protocol.go]: - I guess it does add a little clarity even though redundant (currently).
- Roger that.

File: proxy/api/protocol.go (hunk 2 above)
Comment (dlespiau) [inline proxy/api/protocol.go]: ✓

File: proxy/api/protocol.go (hunk 3 above)
Comment (dlespiau) [inline proxy/api/protocol.go]: ✓

**Comment by jodh-intel** [comment]
lgtm

**Comment by chavafg** [comment]
qa-passed

File: proxy/api/protocol.go (hunk 1 above)
Comment (amshinde) [inline proxy/api/protocol.go]: The shim uses a different limit on the payload, namely the one specified by hyperstart.
https://github.com/01org/cc-oci-runtime/blob/master/shim/shim.h#L61.
We should be consistent here.

Comment (dlespiau) [inline proxy/api/protocol.go]: Except that this limit isn't the max length of a packet is it? just the length of the receive buffer. I'm hoping hyperstart can reconstruct bigger packets then the receive buffer size, but maybe that's too hopeful.

Comment (jcvenegas) [inline proxy/api/protocol.go]: @amshinde could you take a look to @dlespiau  comment ?  If you agreed give a approval :)

Comment (dlespiau) [inline proxy/api/protocol.go]: I was lazy yesterday, but it seems hyperstart will realloc the receive buffer (without any check one the new size, ooops) https://github.com/clearcontainers/hyperstart/blob/0.7.0-clearcontainers/src/init.c#L1193

Comment (amshinde) [inline proxy/api/protocol.go]: that needs to be fixed !

**Comment by amshinde** [comment]
lgtm

**Pull request merged**